### PR TITLE
10226 Force Game Piece Image names to be valid png filenames

### PR DESCRIPTION
**Pull request by BrentEaston**
Newly created GPI's will be forced to be a valid file name containing only 09,a-z,A-Z and _ and ending in .png.

Not much can be done about filenames of existing GPI's as other components will be referencing them. However, if an existing GPI has a valid filename, or is changed to gave a valid filename, then valid filenames will be enforced for that GPI from then on.

### CODE REVIEW

**Comment by uckelman** [comment]
Why allow `$`?

**Comment by BrentEaston** [comment]
Good Point. $ removed.

File: vassal-app/src/main/java/VASSAL/build/module/gamepieceimage/GamePieceImage.java [hunk 1]
```diff
@@ -463,4 +501,167 @@ protected void rebuildInstances() {
...
+
+    private String clean(String string) {
+      final StringBuilder sb = new StringBuilder();
+      for (int i = 0; i < string.length(); i++) {
+        final char c = string.charAt(i);
+        if (c == '_' || (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9')) {
+          sb.append(c);
+        }
+      }
+      return sb.toString();
```
Comment (uckelman) [inline vassal-app/src/main/java/VASSAL/build/module/gamepieceimage/GamePieceImage.java:613]: Could you replace this whole function with `string.replaceAll("\W", "");`?

Comment (BrentEaston) [reply vassal-app/src/main/java/VASSAL/build/module/gamepieceimage/GamePieceImage.java]: Better version.

File: vassal-app/src/main/java/VASSAL/build/module/gamepieceimage/GamePieceImage.java [hunk 2]
```diff
@@ -575,45 +575,34 @@ private boolean isPng(FilterBypass fb) {
...
-    private void fixPng(FilterBypass fb, int pos) throws BadLocationException {
+    private void fixPng(FilterBypass fb, int caretPos) throws BadLocationException {
       if (!isPng(fb)) {
         final String text = getText(fb);
         if (text.endsWith(".")) {
           super.replace(fb, text.length() - 1, 1, PNG_SUFFIX, null);
-          config.setCaretPosition(pos);
+          config.setCaretPosition(caretPos);
         }
         else if (text.toUpperCase().endsWith(".p")) {
```
Comment (jonathanrwatts) [inline vassal-app/src/main/java/VASSAL/build/module/gamepieceimage/GamePieceImage.java:585]: Am I missing something here, or is this test always going to fail (converting to uppercase, then checking for a lowercase letter)?  (Same for the next "else if", too.)

File: vassal-app/src/main/java/VASSAL/build/module/gamepieceimage/GamePieceImage.java [hunk 3]
```diff
@@ -539,8 +539,8 @@ public void setCaretPosition(int pos) {
   }
 
   /**
-   * ImageNameFilter that controls how the user can change the Image Name
-   * If the GPI is still version 0, then no controls. Not we can't force an image
+   * ImageNameFilter that controls how the user can change the Image Name.
+   * If the GPI is still version 0, then no controls. Note we can't force an image
    * name because other components may reference that image.
    * Once the GPI is version 0, then enfore a .png suffix and
```
Comment (jonathanrwatts) [inline vassal-app/src/main/java/VASSAL/build/module/gamepieceimage/GamePieceImage.java:545]: "and" what?

**Pull request merged**